homeassistant/components/recorder/__init__.py

```diff
@@ -313,24 +301,48 @@ def _purge_old_data(self):
 
         purge_before = dt_util.utcnow() - timedelta(days=self.purge_days)
 
-        _LOGGER.info("Purging events created before %s", purge_before)
-        deleted_rows = Session().query(Events).filter(
-            (Events.created < purge_before)).delete(synchronize_session=False)
-        _LOGGER.debug("Deleted %s events", deleted_rows)
+        def _purge_events(session):
+            deleted_rows = session.query(Events) \
+                                  .filter((Events.created < purge_before)) \
+                                  .delete(synchronize_session=False)
+            _LOGGER.debug("Deleted %s events", deleted_rows)
+
+        if self._commit(_purge_events):
+            _LOGGER.info("Purged events created before %s", purge_before)
 
-        _LOGGER.info("Purging states created before %s", purge_before)
-        deleted_rows = Session().query(States).filter(
-            (States.created < purge_before)).delete(synchronize_session=False)
-        _LOGGER.debug("Deleted %s states", deleted_rows)
+        def _purge_states(session):
```

Comment:
I know this is the existing order, but we should first purge the states before purging the events. Otherwise we hit a foreign key constraint.